qiskit/quantum_info/states/statevector.py

```diff
@@ -849,6 +850,10 @@ def _evolve_instruction(statevec, obj, qargs=None):
             return statevec
         if isinstance(obj, Barrier):
             return statevec
+        if isinstance(obj, Initialize):
+            # state is initialized to the statevector in the initialize object
+            statevec._data = np.array(obj.params, dtype=complex)
+            return statevec
```

Comment:
The `dtype` of the params field of `Initialize` isn't always the statevector itself - it can also be things like `"++++"` or whatever.

Reply:
Whoops good point, thanks!

Reply:
`np.asarray` does not copy the object if it's already complex. It is unsafe but it improves performance significantly. (because copy of exponential size data is too expensive)